tokio/src/io/stdio_common.rs

```diff
@@ -25,46 +26,73 @@ where
+        {
+            return call_inner(buf);
+        }
+
+        buf = &buf[..crate::io::blocking::MAX_BUF];
+
+        // Now there are two possibilites.
+        // If caller gave is binary buffer, we **should not** shrink it
+        // anymore, because excessive shrinking hits performance.
+        // If caller gave as binary buffer, we  **must** additionaly
+        // shrink it to strip incomplete char at the end of buffer.
+        // that's why check we will perform now is allowed to have
+        // false-positive.
+
+        // this constant is defined by Unicode standard.
+        const MAX_BYTES_PER_CHAR: usize = 4;
+
+        // Subject for tweaking here
+        const MAGIC_CONST: usize = 8;
+
+        // Now let's look at the first MAX_BYTES_PER_CHAR * MAGIC_CONST bytes.
+        // if they are (possibly incomplete) utf8, then we can be quite sure
+        // that input buffer was utf8.
+
+        let have_to_fix_up = match std::str::from_utf8(&buf[..MAX_BYTES_PER_CHAR * MAGIC_CONST]) {
+            Ok(_) => {
+                // We do not need to shrink this buffer:
+                // we were lucky enough and it didn't broke
+                // during shrinking
+                false
+            }
             Err(err) => {
-                let bad_bytes = buf.len() - err.valid_up_to();
-                // TODO: this is too conservative
-                const MAX_BYTES_PER_CHAR: usize = 8;
-
-                if bad_bytes <= MAX_BYTES_PER_CHAR && err.valid_up_to() > 0 {
-                    // Input data is probably UTF-8, but last char was split
-                    // after trimming.
-                    // let's exclude this character from the buf
-                    &buf[..err.valid_up_to()]
-                } else {
-                    // UTF-8 violation could not be caused by trimming.
-                    // Let's pass buffer to underlying writer as is.
-                    // Why do not we return error here? It is possible
-                    // that stdout is not console. Such streams allow
-                    // non-utf8 data. That's why, let's defer to underlying
-                    // writer and let it return error if needed
-                    buf
-                }
+                let bad_bytes = (MAX_BYTES_PER_CHAR * MAGIC_CONST) - err.valid_up_to();
+                // this must hold for any shrinked utf8 buffer.
+                bad_bytes < MAX_BYTES_PER_CHAR
             }
         };
-        // now pass trimmed input buffer to inner writer
-        Pin::new(&mut self.inner).poll_write(cx, buf)
+
+        if have_to_fix_up {
+            // We must pop several bytes at the end which form incomplete
+            // character. To achieve it, we exploit UTF8 encoding:
+            // for any code point, all bytes except first start with 0b10 prefix.
+            // see https://en.wikipedia.org/wiki/UTF-8#Encoding for details
+            let trailing_incomplete_char_size = buf
```

Comment:
Note for myself: this is just incorrect if buffer is binary in fact (we can pop many bytes or even panic).
Needs fix & test.

Reply:
I think now logic is correct